lib/aee.js

```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+class AsyncEventEmitter {
+  constructor() {
+    this.events = new Map();
+    this.wrappers = new Map();
+  }
+
+  on(name, fn) {
+    const event = this.events.get(name);
+    if (event) event.add(fn);
+    else this.events.set(name, new Set([fn]));
+  }
+
+  once(name, fn) {
+    const wrapper = (...args) => {
+      this.remove(name, wrapper);
+      fn(...args);
+    };
+    this.wrappers.set(fn, wrapper);
+    this.on(name, wrapper);
+  }
+
+  async emit(name, ...args) {
+    const event = this.events.get(name);
+    if (!event) return undefined;
+    const promises = [...event.values()].map(fn => fn(...args));
+    return Promise.all(promises);
+  }
+
+  remove(name, fn) {
+    const event = this.events.get(name);
+    if (!event) return;
+    if (event.has(fn)) {
+      event.delete(fn);
+      return;
+    }
+    const wrapper = this.wrappers.get(fn);
+    if (wrapper) {
+      event.delete(wrapper);
+      if (event.size === 0) this.events.delete(name);
```

Comment:
This should either be in the `if (event.has(fn)) {` branch too or not be here at all IMO.

Reply:
Moved to be executed in both cases